usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2695,6 +2703,282 @@ static CK_RV ccatok_pkey_add_attrs(STDLL_TokData_t * tokdata, TEMPLATE *template
+                 &rule_array_count, rule_array, &clear_key_bit_length,
+                 clear_key_value, &key_name_length, key_name, &user_data_length,
+                 user_data, &token_data_length, NULL, &verb_data_length,
+                 verb_data, &the_key_token_length, the_key_token);
+
+    if (return_code != 0) {
+        TRACE_ERROR("CSNBTKB2 (TOKEN BUILD2) failed with %ld/%ld\n",
+            return_code, reason_code);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Generate the key */
+    memset(rule_array, 0, sizeof(rule_array));
+    memcpy(rule_array, "AES     " "OP      ", 2 * CCA_KEYWORD_SIZE);
+    rule_array_count = 2;
+    clear_key_bit_length = 256;
+    memcpy(key_type_1, "TOKEN   ", CCA_KEYWORD_SIZE);
+    the_key_token_length = sizeof(the_key_token);
+
+    dll_CSNBKGN2(&return_code, &reason_code, NULL, NULL, &rule_array_count,
+                 rule_array, &clear_key_bit_length, key_type_1, key_type_2,
+                 &key_name_1_length, NULL, &key_name_2_length, NULL,
+                 &user_data_1_length, NULL, &user_data_2_length, NULL,
+                 &kek_identifier_1_length, NULL, &kek_identifier_2_length,
+                 NULL, &the_key_token_length, the_key_token,
+                 &the_2nd_key_token_length, NULL);
+
+    if (return_code != 0) {
+        TRACE_ERROR("CSNBKGN2(KEYGEN2) failed with %ld/%ld\n",
+            return_code, reason_code);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Build attribute for secure key token */
+    ret = build_attribute(CKA_IBM_OPAQUE, (CK_BYTE *)&the_key_token,
+                          the_key_token_length, &sec_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("build_attribute CKA_IBM_OPAQUE failed with ret=0x%lx\n", ret);
+        goto done;
+    }
+
+    /*
+     * Create a protected key from this token to obtain the firmware wkvp. When
+     * this function returns ok, we have a 64 byte pkey value: 32 bytes
+     * encrypted key + 32 bytes vp.
+     */
+    ret = ccatok_pkey_skey2pkey(tokdata, sec_attr, &pkey_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("ccatok_pkey_skey2pkey failed with ret=0x%lx\n", ret);
+        goto done;
+    }
+
+    /* Save WKVP in token data */
+    memcpy(&cca_data->pkey_mk_vp, (CK_BYTE *)pkey_attr->pValue + AES_KEY_SIZE_256,
+           PKEY_MK_VP_LENGTH);
+    cca_data->pkey_wrap_supported = 1;
+
+done:
+
+    if (sec_attr)
+        free(sec_attr);
+    if (pkey_attr)
+        free(pkey_attr);
+
+    return ret;
+}
+
+/*
+ * Return true if PKEY_MODE DISABLED is set in the token specific
+ * config file, false otherwise.
+ */
+static CK_BBOOL ccatok_pkey_option_disabled(STDLL_TokData_t *tokdata)
+{
+    struct cca_private_data *cca_data = tokdata->private_data;
+
+    if (cca_data->pkey_mode == PKEY_MODE_DISABLED)
+        return CK_TRUE;
+
+    return CK_FALSE;
+}
+
+/*
+ *                     End of CCA protected key option
```

Comment:
remove this "End of CCA protected key option" comment, too

Reply:
Done.